pkg/kubelet/server/stats/volume_stat_calculator.go

```diff
@@ -113,37 +116,52 @@ func (s *volumeStatCalculator) calcAndStoreStats() {
 		}
 		// Lookup the volume spec and add a 'PVCReference' for volumes that reference a PVC
 		volSpec := volumesSpec[name]
+		var pvcRef *stats.PVCReference
 		if pvcSource := volSpec.PersistentVolumeClaim; pvcSource != nil {
-			pvcRef := stats.PVCReference{
+			pvcRef = &stats.PVCReference{
 				Name:      pvcSource.ClaimName,
 				Namespace: s.pod.GetNamespace(),
 			}
-			fsStats = append(fsStats, s.parsePodVolumeStats(name, &pvcRef, metric))
 			// Set the PVC's prometheus metrics
-			s.setPVCMetrics(&pvcRef, metric)
+			s.setPVCMetrics(pvcRef, metric)
+		}
+		volumeStats, isEphemeral := s.parsePodVolumeStats(name, pvcRef, metric, volSpec)
+		if isEphemeral {
+			ephemeralStats = append(ephemeralStats, volumeStats)
 		} else {
-			fsStats = append(fsStats, s.parsePodVolumeStats(name, nil, metric))
+			persistentStats = append(persistentStats, volumeStats)
 		}
+
 	}
 
 	// Store the new stats
-	s.latest.Store(PodVolumeStats{Volumes: fsStats})
+	s.latest.Store(PodVolumeStats{EphemeralVolumes: ephemeralStats,
+		PersistentVolumes: persistentStats})
 }
 
 // parsePodVolumeStats converts (internal) volume.Metrics to (external) stats.VolumeStats structures
-func (s *volumeStatCalculator) parsePodVolumeStats(podName string, pvcRef *stats.PVCReference, metric *volume.Metrics) stats.VolumeStats {
+func (s *volumeStatCalculator) parsePodVolumeStats(podName string, pvcRef *stats.PVCReference, metric *volume.Metrics, volSpec v1.Volume) (stats.VolumeStats, bool) {
 	available := uint64(metric.Available.Value())
 	capacity := uint64(metric.Capacity.Value())
 	used := uint64(metric.Used.Value())
 	inodes := uint64(metric.Inodes.Value())
 	inodesFree := uint64(metric.InodesFree.Value())
 	inodesUsed := uint64(metric.InodesUsed.Value())
+
 	return stats.VolumeStats{
 		Name:   podName,
 		PVCRef: pvcRef,
 		FsStats: stats.FsStats{Time: metric.Time, AvailableBytes: &available, CapacityBytes: &capacity,
 			UsedBytes: &used, Inodes: &inodes, InodesFree: &inodesFree, InodesUsed: &inodesUsed},
+	}, isVolumeEphemeral(volSpec)
```

Comment:
I'm not sure it makes sense to plumb the call to isVolumeEphemeral(volSpec) through the parsePodVolumeStats function.  Can we just call it separately?  Then the above is just `if isVolumeEphemeral(volSpec) {`

Reply:
fixed